enhancements/kubernetes-nmstate-integration.md

```diff
@@ -0,0 +1,165 @@
+[SR-IOV](https://github.com/hustcat/sriov-cni), and more...
+
+However, in all of these cases, the node must have the networks setup before the
+pod is scheduled. Setting up the networks in a dynamic and heterogenous cluster,
+with dynamic networking requirements, is a challenge by itself - and this is
+what this project is addressing.
+
+### Goals
+
+- Deploy kubernetes-nmstate as part of openshift
+
+### Non-Goals
+
+- Replace SRIOV operator
+
+## Proposal
+
+A new kubernetes-nmstate operator Deployment is deployed to the cluster as part of the the OpenShift
+installation, via CVO.
+
+The operator has a single CRD, called `NMstate`. When a custom resource of kind `NMstate`, with 
+a name of 'nmstate' is created, it will then create the CRDs for the kubernetes-nmstate handler.
+Then, the namespace, RBAC, and finally the DaemonSet are applied. Custom resources of kind `NMstate`
+with other names will be ignored.
+
+A new kubernetes-nmstate handler DaemonSet is deployed in the cluster by the operator.
+This DaemonSet contains nmstate package and interacts with the NetworkManager
+on the host by mounting the related dbus. The project contains three
+Custom Resource Definitions, `NodeNetworkState`, `NodeNetworkConfigurationPolicy` and
+`NodeNetworkConfigurationEnactment`. `NodeNetworkState` objects are created per each
+node in the cluster and can be used to report available interfaces and network configuration.
+These objects are created by kubernetes-nmstate and must not be touched by a user.
+`NodeNetworkConfigurationPolicy` objects can be used to specify desired
+networking state per node or set of nodes. It uses API similar to `NodeNetworkState`. 
+`NodeNetworkConfigurationEnactment` objects will be created per each node, per each matching
+`NodeNetworkConfigurationPolicy`.
+
+kubernetes-nmstate DaemonSet creates a custom resource of `NodeNetworkState` type per each node and
+updates the network topology from each OpenShift node.
+
+User configures host network changes and apply a policy in `NodeNetworkConfigurationPolicy` custom
+resource. Network topology is configured via `desiredState` section in `NodeNetworkConfigurationPolicy`.
+Multiple `NodeNetworkConfigurationPolicy` custom resources can be created.
+
+Upon receiving a notification event of `NodeNetworkState` update,
+kubernetes-nmstate Daemon verify the correctness of `NodeNetworkState` custom resource and
+apply the selected profile to the specific node.
+
+`NodeNetworkConfigurationEnactment` objects is a read-only object that represents Policy exectuion
+per each matching Node. It will expose configuration status per each Node.
+
+A new container image (kubernetes-nmstate-handler) will be created.
+
+A new container image (kubernetes-nmstate-operator) will be created. However, the operator
+and the handler will co-exist in the same upstream repo.
+
+An upstream API group of nmstate.io is currently used.
+
+### User Stories
+
```

Comment:
Meta-criticisim - these aren't user stories. User stories look like this:

**"As a (person), I'd like to do (high level goal)."** Start with a goal, then show how your proposed solution fulfills this story. **"With (feature x, y, and z), I can accomplish this.""**

It's important that user stories are written **without the proposed solution in mind**. Ideally they come "first" in the design process. Their role is to identify real needs, rather than extant features. That way you can be sure you're designing a solution for problems, not the other way around.

An example:
> I a cluster administrator with a dedicated storage network, because performance for me is critical. I need to make sure that all storage traffic uses the dedicated network, and I need to be able to schedule my Gluster pods on nodes with 10GB interfaces.
> I would configure my routing tables with k8s-mnstate and (filling in the details here). The existing functionality doesn't work for me because (of all these reasons).

Make sense?

Reply:
I strongly agree with @squeed .

The most important argument this enhancement needs to make is a clear argument for why this capability is required on all OpenShift clusters by default.  Nothing in this enhancement resonates in its current form explaining why this is needed univerally in OpenShift rather than as an advanced optional add-on component delivered via OLM.